crates/sovereign-sdk/module-system/sov-state/src/witness.rs

```diff
@@ -17,28 +20,54 @@ use sov_modules_core::Witness;
 /// assert_eq!(witness.get_hint::<u64>(), 1u64);
 /// assert_eq!(witness.get_hint::<u64>(), 2u64);
 /// ```
-#[derive(Default, BorshDeserialize, BorshSerialize, Debug, Serialize, Deserialize)]
+#[derive(Default, Debug, Serialize, Deserialize)]
 pub struct ArrayWitness {
-    next_idx: usize,
-    hints: Vec<Vec<u8>>,
+    next_idx: AtomicUsize,
+    hints: Mutex<Vec<Vec<u8>>>,
 }
 
 impl Witness for ArrayWitness {
-    fn add_hint<T: BorshSerialize>(&mut self, hint: T) {
-        self.hints.push(borsh::to_vec(&hint).unwrap())
+    fn add_hint<T: BorshSerialize>(&self, hint: T) {
+        self.hints
+            .lock()
+            .unwrap()
+            .push(borsh::to_vec(&hint).unwrap());
     }
 
-    fn get_hint<T: BorshDeserialize>(&mut self) -> T {
-        let idx = self.next_idx;
-        self.next_idx += 1;
-        T::deserialize_reader(&mut std::io::Cursor::new(&self.hints[idx]))
+    fn get_hint<T: BorshDeserialize>(&self) -> T {
+        let idx = self
+            .next_idx
+            .fetch_add(1, std::sync::atomic::Ordering::SeqCst);
+        let hints_lock = self.hints.lock().unwrap();
+        T::deserialize(&mut hints_lock[idx].as_slice())
             .expect("Hint deserialization should never fail")
     }
 
-    fn merge(&mut self, rhs: &mut Self) {
-        let rhs_next_idx = rhs.next_idx;
-        let lhs_hints_lock = &mut self.hints;
-        let rhs_hints_lock = &mut rhs.hints;
-        lhs_hints_lock.extend(rhs_hints_lock.drain(rhs_next_idx..))
+    fn merge(&self, rhs: &Self) {
+        let rhs_next_idx = rhs.next_idx.load(std::sync::atomic::Ordering::SeqCst);
+        let mut lhs_hints_lock = self.hints.lock().unwrap();
+        let mut rhs_hints_lock = rhs.hints.lock().unwrap();
```

Comment:
This is sort of scary. `self.hints` can be locked, and in a completely different operation, the order of locking could be first rhs, and then self, which would cause deadlock. I know there is no such case, but I think it would be safer if we:
1. Create empty vec: merged
2. lock lhs and extend merged vec with lhs
3. lock rhs and extend merged with rhs

This way we won't hold 2 resources locked at the same time, much less risk of deadlock of any sort for possible future iterations

Reply:
You forgot the 4th step, lock lhs again and set merged into it.

Which is pretty much what the code is doing. 

> self.hints can be locked, and in a completely different operation, the order of locking could be first rhs, and then self, which would cause deadlock.

I don't think we need to consider this. 

IFF we ever introduce new operations that would lock in the wrong ordering, we'll know and we'll resolve. This PR doesn't need to address imaginary issues right now.

Reply:
I am opposing your approach because of lock / release, lock / release, lock / release cycle (lhs needs to be locked twice)

Reply:
You are right, locking lhs 2 times is not good. Then I modify my suggestion to:

1. Lock rhs and clone it
2. Lock lhs and extend it with cloned rhs

Everything is locked just once, but not at the same time.

Reply:
Done